### PR TITLE
feat: support webpack-dev-server v5 in @cypress/webpack-dev-server

### LINKED ISSUE #29305: Support webpack-dev-server v5 in @cypress/webpack-dev-server
### What would you like?

I would like to @cypress/webpack-dev-server to support webpack-dev-server v5

### Why is this needed?

new version of webpack-dev-server is available, which projects like angular v18 are using and others likely will use as well

### Other

To make this not a breaking change to cypress nad @cypress/webpack-dev-server, webpack-dev-server version 4 still needs to be the default/shipped version, but webpack-dev-server 5 should be able to be sourced without error

### PR DESCRIPTION
**Pull request by AtofStryker**
- Closes #29305

### Additional details

This PR introduces support for webpack-dev-server v5 inside @cypress/webpack-dev-server.

To avoid a breaking change, @cypress/webpack-dev-server still ships with wds v4 by default. This PR allows v5 to be discovered and supported if a user has wds v5 installed in their project.

The configuration is almost identical to that of wds v4, but I elected to keep both separate as feature configuration could evolve with v5 and not with v4, allowing us flexibility by keeping these options separate.

I have also added some system tests to verify things work as expected with wds5 and webpack 5. wds 5 does not work with webpack 4, so I added a custom error message inside of @cypress/webpack-dev-server to alert the user if they are using wds5 and webpack 4, as the error thrown under the hood is not graceful nor obvious.

I have also added this as a `feat` for cypress as this `@cypress/webpack-dev-server` does ship with cypress by default.

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by AtofStryker** [comment]
The changes in the network package have to do with types being installed in `@cypress/webpack-dev-server` which now installs the forge types and causes conflicts. Since we don't use the types in the network package and they are not compatible with the `@types/node-forge` package, I opted for a require to ignore the types

File: npm/webpack-dev-server/package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "@types/proxyquire": "^1.3.28",
     "@types/speed-measure-webpack-plugin": "^1.3.4",
     "@types/webpack-dev-server-3": "npm:@types/webpack-dev-server@^3",
+    "@types/webpack-dev-server-4": "npm:@types/webpack-dev-server@^4",
```
Comment (jordanpowell88) [inline npm/webpack-dev-server/package.json]: If case you didn't realize. It looks like https://www.npmjs.com/package/@types/webpack-dev-server has been deprecated.

File: npm/webpack-dev-server/src/helpers/sourceRelativeWebpackModules.ts [hunk 2]
```diff
@@ -180,11 +187,10 @@ export function sourceWebpack (config: WebpackDevServerConfig, framework: Source
 
 // Source the webpack-dev-server module from the provided framework or projectRoot.
 // If none is found, we fallback to the version bundled with this package.
-export function sourceWebpackDevServer (config: WebpackDevServerConfig, framework?: SourcedDependency | null): SourcedWebpackDevServer {
+export function sourceWebpackDevServer (config: WebpackDevServerConfig, webpackMajorVersion: 4 | 5, framework?: SourcedDependency | null): SourcedWebpackDevServer {
```
Comment (jordanpowell88) [inline npm/webpack-dev-server/src/helpers/sourceRelativeWebpackModules.ts:190]: Is 3 the version that is the fallback and why it isn't part of this `webpackMajorVersion` union here?

Comment (AtofStryker) [reply npm/webpack-dev-server/src/helpers/sourceRelativeWebpackModules.ts:190]: luckily for us we don't support webpack v3, only v4 and v5. But with this PR we would support webpack-dev-server v3, v4, and v5.

the package currently ships with webpack-dev-server v4 but will source whatever the user has installed

File: npm/webpack-dev-server/package.json (hunk 1 above)
Comment (AtofStryker) [reply npm/webpack-dev-server/package.json]: oh nice It looks like it ships with 4 by default. I don't think I am using them anyway so we can remove them and maybe revert the changes to the network

**Comment by jennifer-shehane** [comment]
@AtofStryker Should we regard this as closed when this goes in? Or do we need some extra tests to verify on top of this? https://github.com/cypress-io/cypress/issues/29309

**Comment by AtofStryker** [comment]
> @AtofStryker Should we regard this as closed when this goes in? Or do we need some extra tests to verify on top of this? #29309

This should be marked as closed once it goes in since the system tests are verifying that the integration works, unless there are new features we need to check for that we don't know about. The only thing we need to do on top of this is would be #28926 but there isnt a strict timeline on that since it is a breaking change.

**Comment by AtofStryker** [comment]
> @AtofStryker Should we regard this as closed when this goes in? Or do we need some extra tests to verify on top of this? #29309

we need to officially support Angular 18 but this gets us a step closer. I was talking with @jordanpowell88 about this today and sounds like the scope of work isn't massive? i.e. we need some tests on top of it. The ones I know about are the system tests

File: npm/webpack-dev-server/package.json (hunk 1 above)
Comment (AtofStryker) [reply npm/webpack-dev-server/package.json]: looks like that worked. removed in 880cfa1256

**Comment by jennifer-shehane** [comment]
@AtofStryker This should also resolve these github actions failing outright: 
- https://github.com/cypress-io/ct-automation/actions/runs/8639464228/job/23685817659
- https://github.com/cypress-io/ct-automation/actions/runs/8639464228/job/23685818766

**Comment by AtofStryker** [comment]
> @AtofStryker This should also resolve these github actions failing outright:
> 
> * https://github.com/cypress-io/ct-automation/actions/runs/8639464228/job/23685817659
> * https://github.com/cypress-io/ct-automation/actions/runs/8639464228/job/23685818766

yes that it should!

File: cli/CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +16,7 @@ _Released 4/11/2024_
 - Fixed an issue where asserts with custom messages weren't displaying properly. Fixes [#29167](https://github.com/cypress-io/cypress/issues/29167).
 - Fixed and issue where Cypress launch arguments were not being escaped correctly with multiple values inside quotes. Fixes [#27454](https://github.com/cypress-io/cypress/issues/27454).
 
+
```
Comment (AtofStryker) [inline cli/CHANGELOG.md]:
```suggestion
```

**Review by cacieprins — APPROVED** [review]

File: npm/webpack-dev-server/src/helpers/sourceRelativeWebpackModules.ts [hunk 4]
```diff
@@ -208,9 +214,18 @@ export function sourceWebpackDevServer (config: WebpackDevServerConfig, framewor
...
+  webpackDevServer.majorVersion = getMajorVersion(webpackDevServer.packageJson, [3, 4, 5])
 
   debug('WebpackDevServer: Successfully sourced webpack-dev-server - %o', webpackDevServer)
+  if (webpackMajorVersion < 5 && webpackDevServer.majorVersion === 5) {
+    const json = webpackDevServer.packageJson
+
+    throw new CypressWebpackDevServerError(
+      `Incompatible major versions of webpack and webpack-dev-server! ` +
+      `webpack-dev-server major version ${webpackDevServer.majorVersion} only works with major versions of webpack ${webpackMajorVersion} - saw webpack-dev-server version ${json.version}. ` +
+      `If using webpack major version 4, please install webpack-dev-server version 4 to be used with @cypress/webpack-dev-server or upgrade to webpack 5.`,
```
Comment (mschile) [inline npm/webpack-dev-server/src/helpers/sourceRelativeWebpackModules.ts]:
```suggestion
      `Incompatible major versions of webpack and webpack-dev-server!
      webpack-dev-server major version ${webpackDevServer.majorVersion} only works with major versions of webpack ${webpackMajorVersion} - saw webpack-dev-server version ${json.version}.
      If using webpack major version 4, please install webpack-dev-server version 4 to be used with @cypress/webpack-dev-server or upgrade to webpack 5.`,
```

File: npm/webpack-dev-server/tsconfig.json [hunk 5]
```diff
@@ -45,9 +45,9 @@
     // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
     // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
     "esModuleInterop": true,
+    "skipLibCheck": true,
     /** Allows us to strip internal types sourced from webpack */
-    "stripInternal": true,
-    "importsNotUsedAsValues": "error"
```
Comment (mschile) [inline npm/webpack-dev-server/tsconfig.json:50]: Why was this removed?

File: npm/webpack-dev-server/package.json [hunk 6]
```diff
@@ -42,7 +42,8 @@
     "ts-node": "^10.9.2",
     "webpack": "npm:webpack@^5",
     "webpack-4": "npm:webpack@^4",
-    "webpack-dev-server-3": "npm:webpack-dev-server@^3"
+    "webpack-dev-server-3": "npm:webpack-dev-server@^3",
+    "webpack-dev-server-5": "npm:webpack-dev-server@^5"
```
Comment (mschile) [inline npm/webpack-dev-server/package.json:46]: Out of curiosity, why isn't `npm:webpack-dev-server@v4` listed?

**Review by mschile — APPROVED** [review]
Added a few comments but looks good!

File: npm/webpack-dev-server/package.json (hunk 6 above)
Comment (AtofStryker) [reply npm/webpack-dev-server/package.json:46]: because its installed under `webpack-dev-server` since its the default that ships with `@cypress/webpack-dev-server`. Eventually we want to bump it to 5, but that is a breaking change

File: npm/webpack-dev-server/tsconfig.json (hunk 5 above)
Comment (AtofStryker) [reply npm/webpack-dev-server/tsconfig.json:50]: I was getting deprecation warnings and I think I forgot to migrate it to the new flag https://www.typescriptlang.org/tsconfig#preserveValueImports. Ill try adding it back in and see whats up
